docs/junit-integration.md

```diff
@@ -0,0 +1,91 @@
+# JUnit Integration Implementation
+
+Jazzer's JUnit integration starts from
+the [`FuzzTest`](../src/main/java/com/code_intelligence/jazzer/junit/FuzzTest.java) annotation. As mentioned in the
+annotation's javadoc, our integration runs in one of two modes: fuzzing and regression. Fuzzing mode will generate new
+inputs to feed into the tests to find new issues and regression mode will run the tests against previous findings, no
+fuzzing is done. The main entrypoints for the actual integration code are found in two of the annotations
+on `FuzzTest`: `@ArgumentsSource(FuzzTestArgumentsProvider.class)` and `@ExtendsWith(FuzzTestExtensions.class)`.
+
+Because these same files and functions are involved in two mostly separate sets of functionality, this will look at the
+flow of the different methods involved in integrating with JUnit in fuzzing mode (when `JAZZER_FUZZ` is set to any
+non-empty value) and in regression mode (when `JAZZER_FUZZ` is not set) separately.
+
+# Fuzzing Flow
+
+## `evaluateExecutionCondition`
+
+It appears that, when running an entire class of tests, JUnit will first call this function on all tests marked
+with `@FuzzTest` looking for at least one method to return that it should be run. If a test class has multiple fuzz tests, it
+will short circuit at the first that says it's enabled. In fuzzing mode, Jazzer will only run a single test so this will
+save this first test and later calls to this function will return enabled for only this test.
+
+## `provideArguments`
+
+This will configure the fuzzing agent to set up code instrumentation, instantiate a `FuzzTestExecutor` and put it into
+JUnit's `extensionContext`, then create a stream of a single empty argument set. As the comment mentions, this is so
+that JUnit will actually execute the test but the argument will not be used.
+
+## `evaluateExecutionCondition`
+
+This will be called again for each fuzz test that might be run i.e. it does not short circuit. As
+stated above, `FuzzTestExtensions` will cache the first test to be checked and only allow that to run so all other tests
+will return disabled.
+
+## `interceptTestTemplateMethod`
+
+This will call `invocation.skip()` which prevents invoking the test function with the default set of
+arguments `provideArguments` created. It will instead extract the `FuzzTestExecutor` instance from
+the `extensionContext` and then calls `FuzzTestExecutor#execute` which creates a `FuzzTargetRunner` to run the actual
+fuzzing.
+
+Crashes are saved in `resources/<package>/<test file name>Inputs/<test method name>` and results that are interesting to
+libfuzzer are saved in `.cifuzz-corpus` though where that's done is a bit of a mystery to me.
+
+# Regression Flow
+
+## `evaluateExecutionCondition`
+
+Similar to the first execution of this in fuzzing mode, this will look for the first fuzz test to return that it's
```

Comment:
In regression test mode we should be running all fuzz tests (or have a bug somewhere). Since regression test doesn't run libFuzzer, we don't have to worry about its global state.

Reply:
I had the usage of this method by JUnit wrong. It's called once on the test itself to determine if the test should be run at all and if so then it will be called again for each input it receives from `provideArguments`. So it goes:

1. `evaluateExecutionCondition` to determine if this test is enabled at all
2. `provideArguments` to get the set of arguments that this test should be called with
3. For each argument set from `provideArguments`, `evaluateExecutionCondition` is called again to check if that particular argument set is enabled.

In fuzzing mode, that would go:
```
fuzzTest1 - enabled
fuzzTest1 Fuzzing... - enabled
fuzzTest2 - disabled
fuzzTest3 - disabled
```
and in regression:
```
fuzzTest1 - enabled
fuzzTest1 <empty input> - enabled
fuzzTest1 crash-1 - enabled
fuzzTest2 - enabled
fuzzTest2 <empty input> - enabled
fuzzTest2 crash-2 - enabled
...
```

Which meant that it looked like it was short circuiting because it called `evaluateExecutionCondition` until it found an enabled test, then called `provideArguments`, then `evaluateExecutionCondition` again but before I wasn't printing enough context to know why those additional calls were happening.